samples/drivers/flash_shell/src/main.c

```diff
@@ -227,6 +248,143 @@ static int do_write(const struct shell *shell, off_t offset, uint8_t *buf,
+	}
+
+	if (0 == size_before && 0 == size_after) {
+		/* Aligned write */
+		flash_erase(flash_device, offset, len);
+		flash_write(flash_device, offset, buf, len);
+		ret = flash_write_protection_set(flash_device, true);
+		if (ret) {
+			PR_ERROR(shell, "Failed to enable flash protection (err: %d)."
+					"\n", ret);
+		}
+		return ret;
+	}
+
+	before_data = k_malloc(page_size);
+	after_data = k_malloc(page_size);
+
+	if (!before_data || !after_data) {
+		PR_ERROR(shell, "No heap memory for flash manipulation\n");
+		ret = -ENOMEM;
+		read_back = false;
+		goto free_buffers;
+	}
+
+	/* Stash useful data from the pages that will be affected. */
+	if (single_page_write) {
+		/* Read old data before new data is written. */
+		if (size_before) {
+			flash_read(flash_device, start_page, before_data, size_before);
+		}
+		/* Fill the with new data. */
+		memcpy(before_data + size_before, buf, len);
+		/* Fill the last part of old data. */
+		if (size_after) {
+			flash_read(flash_device, offset + len,
+					before_data + size_before + len,
+					size_after);
+		}
+	} else {
+		/* Multipage write, different start and end pages. */
+		if (size_before) {
+			flash_read(flash_device, start_page, before_data,
+					size_before);
+			/* Fill the rest with new data. */
+			memcpy(before_data + size_before, buf,
+			       page_size - size_before);
+		}
+		if (size_after) {
+			/* Copy ending part of new data. */
+			memcpy((void *)after_data,
+			       (void *)(buf + len -
+				   ((len + size_before) % page_size)),
+			       page_size - size_after);
+			/* Copy ending part of flash page. */
+			flash_read(flash_device, offset + len,
+					after_data + (page_size - size_after),
+					size_after);
+		}
+	}
+
+	/* Erase all the pages that overlap with new data. */
+	flash_erase(flash_device, start_page, aligned_size);
+
+	/* Write stashed and new data. */
+	if (single_page_write || size_before > 0) {
+		/* Write first page if available. */
+		flash_write(flash_device, start_page, before_data,
+				 page_size);
+	}
+	if (!single_page_write) {
+		size_t middle_data_len = aligned_size;
+		off_t middle_page_start = start_page;
+		off_t data_offset = (off_t)buf;
+
+		/* Write the middle bit if available */
+		if (size_before > 0) {
+			middle_page_start += page_size;
+			middle_data_len -= page_size;
+			data_offset += (page_size - size_before);
+		}
+		if (size_after > 0) {
+			middle_data_len -= page_size;
+		}
+		if (middle_data_len > 0) {
+			flash_write(flash_device, middle_page_start,
+					 (const void *)data_offset,
+					 middle_data_len);
+		}
+
+		/* Write the last page if needed. */
+		if (size_after > 0) {
+			flash_write(flash_device, last_page, after_data,
+					 page_size);
+		}
+	}
+
+free_buffers:
+	k_free(before_data);
+	k_free(after_data);
+
+	ret = flash_write_protection_set(flash_device, true);
+	if (ret) {
+		PR_ERROR(shell, "Failed to enable flash protection (err: %d)."
+				"\n", ret);
+		return ret;
+	}
+
+	if (read_back) {
```

Comment:
This does do the read back, even if something previously failed. Maybe the cleanup should be after this?

Reply:
Okay, will move the cleanup at the end of the function.